### ISSUE TITLE
[QUESTION] Are bounds checking and floating-point exceptions included in CMake build options?

### ISSUE THREAD

**Issue by msulprizio** [geoschem/geos-chem#303]
I am trying to chase down what I think is an array-out-of-bounds error, so I've turned on the debug option in CMake. However, it isn't returning an information which led me to wonder if the debug option actually turns on checking bounds. I don't see any reference to bounds checking or floating point exceptions in the CMake files. We use both of those options regularly in debugging GEOS-Chem (that is the `BOUNDS=y FPE=y` switches in GNU Make) so we should also include them in CMake. In `Makefile_Header.mk`:

```
  # Turn on checking for floating-point exceptions                                        
  # These are approximately equivalent to -fpe0 -ftrapuv in IFORT                         
  # NOTE: GNU Fortran 4.4.7 does not allow for -finit-real-snan, so                       
  # we will only add this flag for versions newer than 4.4.7                              
  REGEXP             :=(^[Yy]|^[Yy][Ee][Ss])
  ifeq ($(shell [[ "$(FPE)" =~ $(REGEXP) ]] && echo true),true)
    FFLAGS           += -ffpe-trap=invalid,zero,overflow
    ifeq ($(NEWER_THAN_447),1)
      FFLAGS           += -finit-real=snan
    endif
  endif
  ifeq ($(shell [[ "$(FPEX)" =~ $(REGEXP) ]] && echo true),true)
    FFLAGS           += -ffpe-trap=invalid,zero,overflow
    ifeq ($(NEWER_THAN_447),1)
      FFLAGS           += -finit-real=snan
    endif
  endif
  # Add option for "array out of bounds" checking                                         
  REGEXP             := (^[Yy]|^[Yy][Ee][Ss])
  ifeq ($(shell [[ "$(BOUNDS)" =~ $(REGEXP) ]] && echo true),true)
    FFLAGS           += -fbounds-check
  endif
```

@LiamBindle can you confirm that (1) these options aren't already included in GEOS-Chem Classic's CMake options so far? and (2) advise on the best way to add this capability? If I understand correctly, I think we need to add to `GC-DefaultCompilerOptions.cmake`. Does it makes sense to include these in the debug option or as separate options?

**Comment by LiamBindle** [comment]
@msulprizio Yeah, you're correct and `GC-DefaultCompilerOptions.cmake` is the right place to add them. I think it makes sense to add these--initially I omitted most of the debug flags with the idea that we'd add them on a need-to-use basis (to avoid carrying forward any old unsed flags). Feel free to update these to whatever you think is appropriate.

ifort's debug options for debug mode are set here:
https://github.com/geoschem/geos-chem/blob/e3b3b6570af07b1cde3a6d25b143451da25dbb9f/CMakeScripts/GC-DefaultCompilerOptions.cmake#L9
and gfortran's options for debug mode are set here:
https://github.com/geoschem/geos-chem/blob/e3b3b6570af07b1cde3a6d25b143451da25dbb9f/CMakeScripts/GC-DefaultCompilerOptions.cmake#L18

**Comment by LiamBindle** [comment]
Oh, one more thing that I think might be relevent. Just so you know, you can set extra compiler options for your build directory by setting `CMAKE_Fortran_FLAGS` (see [here](http://wiki.seas.harvard.edu/geos-chem/index.php/CMake_FAQ#How_do_I_build_with_custom_compiler_flags.3F)). I think the flags that you're talking about should be added to the DEBUG flags, but I just wanted to mention this feature to increase awareness of it. This feature is useful for building GEOS-Chem with uncommon compiler options, like compiler options for using TAU.

**Comment by msulprizio** [comment]
Thanks, @LiamBindle. This seems straightforward, so I will add bounds checking and FPE to the debug option in [12.8.1](https://github.com/geoschem/geos-chem/milestone/12).

**Comment by lizziel** [comment]
Also heads up that file `geos-chem/CMakeScripts/GC-DefaultCompilerOptions.cmake `is deleted starting 13.0.0. Once this update goes into 12.8.1 we should also port it to dev/13.0.0.

**Comment by msulprizio** [comment]
This is now resolved in bc0c69fd885c7eae58bc8aea58e7423b656ff0be and will be included in 12.8.1. 

The same fix has also been pushed to `dev/13.0.0` in commit 2d0a7cb112be345d4a13b807baac893583bdfa23.
